src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+                            return false;
+                        }
+                    } else if (ioPort.type === 'vec2' && !ioPort.name.startsWith('OUT_') ) {
+                        if (!(ioPort.valueX != undefined && typeof(ioPort.valueX) === 'number' &&
+                              ioPort.valueY != undefined && typeof(ioPort.valueY) === 'number' )) {
+                            return false;
+                        }
+                    } else if (ioPort.type === 'vec3' && !ioPort.name.startsWith('OUT_') ) {
+                        if (!(ioPort.valueX != undefined && typeof(ioPort.valueX) === 'number' &&
+                              ioPort.valueY != undefined && typeof(ioPort.valueY) === 'number' &&
+                              ioPort.valueZ != undefined && typeof(ioPort.valueZ) === 'number' )) {
+                            return false;
+                        }
+                    } else if (ioPort.type === 'vec4' && !ioPort.name.startsWith('OUT_') ) {
+                        if (!(ioPort.valueX != undefined && typeof(ioPort.valueX) === 'number' &&
+                              ioPort.valueY != undefined && typeof(ioPort.valueY) === 'number' &&
+                              ioPort.valueZ != undefined && typeof(ioPort.valueZ) === 'number' &&
+                              ioPort.valueW != undefined && typeof(ioPort.valueW) === 'number' )) {
+                            return false;
+                        }
+                    }
+                } else {
+                    return false;
+                }
+            }
+
+            if (this.graphData.customFuncGlsl) {
+                if (typeof(this.graphData.customFuncGlsl) === 'number') {
+                    return false;
+                }
+            } else {
+                for (i = 0; i < this.graphData.connections.length; i++) {
+                    if (this.graphData.connections[i]) {
+                        if (!(this.graphData.connections[i].srcVarName && this.graphData.connections[i].dstVarName && (this.graphData.connections[i].srcIndex >= 0 || this.graphData.connections[i].dstIndex >= 0 ))) {
+                            return false;
+                        }
+                    } else {
+                        return false;
+                    }
+                }
+
+                // infinite recursion protection
+                // if no graphRootTime set, set it to current time
+                if (!graphRootTime) {
+                    graphRootTime = new Date().getTime();
+                    this._tmpRootTime = graphRootTime;
+                }
+                for (i = 0; i < this.graphData.subGraphs.length; i++) {
+                    if (this.graphData.subGraphs[i] && this.graphData.subGraphs[i] instanceof ShaderGraphNode && this.graphData.subGraphs[i].name) {
+                        if (this.graphData.subGraphs[i]._tmpRootTime === graphRootTime) {

```

Comment:
it seem that the time is used as unique Id here? If yes, I think we should use some variable we keep incrementing instead, as the time has low precision (2ms on some browsers) and so will not give us unique id at all times.

Reply:
okay - I'll switch it to a counter

Reply:
done - thanks!